src/System/Random/Internal.hs

```diff
@@ -499,9 +502,50 @@ class Uniform a where
   -- | Generates a value uniformly distributed over all possible values of that
   -- type.
   --
+  -- There is also a default implementation for finitely-inhabited types.
+  --
+  -- >>> :set -XDeriveGeneric -XDeriveAnyClass
+  -- >>> import GHC.Generics (Generic)
+  -- >>> import System.Random.Stateful
+  -- >>> data MyBool = MyTrue | MyFalse deriving (Show, Generic, Finite, Uniform)
+  -- >>> data Action = Code MyBool | Eat (Maybe Bool) | Sleep deriving (Show, Generic, Finite, Uniform)
+  -- >>> gen <- newIOGenM (mkStdGen 42)
+  -- >>> uniformListM 10 gen :: IO [Action]
+  -- [Code MyTrue,Code MyTrue,Eat Nothing,Code MyFalse,Eat (Just False),Eat (Just True),Eat Nothing,Eat (Just False),Sleep,Code MyFalse]
+  --
   -- @since 1.2.0
   uniformM :: StatefulGen g m => g -> m a
 
+  default uniformM :: (StatefulGen g m, Generic a, GUniform (Rep a)) => g -> m a
+  uniformM = fmap to . guniformM
+
+class GUniform f where
+  guniformM :: StatefulGen g m => g -> m (f a)
+
+instance GUniform f => GUniform (M1 i c f) where
+  guniformM = fmap M1 . guniformM
+
+instance Uniform a => GUniform (K1 i a) where
+  guniformM = fmap K1 . uniformM
+
+instance GUniform U1 where
+  guniformM _ = return U1
+
+instance (GUniform f, GUniform g) => GUniform (f :*: g) where
+  guniformM g = (:*:) <$> guniformM g <*> guniformM g
```

Comment:
This definition appears to be problematic: `Generic` representation fails to fuse because of monadic binds in the midst of `(:*:)`. Even for 
```haskell
{-# LANGUAGE DeriveGeneric, DeriveAnyClass #-}
data Tuple = Triple !Word !Word deriving (Generic, Uniform)
```
generated Core is unwieldy:
```haskell
-- RHS size: {terms: 62, types: 742, coercions: 210, joins: 0/5}
$fUniformTuple_$cuniformM
  :: forall g (m :: * -> *). StatefulGen g m => g -> m Tuple
```

`GFinite`-based version avoids this shortcoming by executing effectful computation exactly ones.

Reply:
When we derive `Uniform` via `GFinite`, cardinality is usually statically known, so we immediately generate a random `m Integer`, then create a generic representation `Integer -> Rep a` and convert it to a concrete data `Rep a -> a`. Since both latter steps are pure, GHC is happy to inline everything and avoid actual construction of `Rep a` altogether.

By using `GUniform` we construct a random `m (Rep a)`, then convert it to a concrete representation `Rep a -> a`. Because of a polymorphic monad `m`, GHC cannot really inline stuff through monadic binds and we end up constructing `Rep a` in full. 

I guess that in client's code, which works in a fixed monomorphic `m`, the definition of `uniformM` would inline specialized and actually avoid allocating `Rep a`. But this is a fragile assumption, so I decided to export `uniformViaFiniteM` to be on the safe side.

Reply:
@Bodigrim I made a quick test with converting to continuations. General idea is to work in `ContT m` instead of `m`

```haskell
class GUniform f where
  guniformM :: StatefulGen g m => g -> ContT r m (f a)
```

Since `ContT` monad instance doesn't use bind of underlying monad GHC is free to inline and otherwise specialize. And it's able to remove all traces of Generic from generated instance

```
from: -- RHS size: {terms: 61, types: 589, coercions: 246, joins: 0/8}
to:   -- RHS size: {terms: 17, types: 76, coercions: 0, joins: 0/0}
+     -- RHS size: {terms: 30, types: 51, coercions: 0, joins: 0/1}

```

Reply:
Awesome! What is `+`?

Reply:
GHC split definition in two functions. That's count for both